core/pkg/artifacts/saver.go

```diff
@@ -144,99 +146,157 @@ func (as *ArtifactSaver) uploadFiles(artifactID string, manifest *Manifest, mani
+		// Still, as long as more than half of them succeed in each iteration this will
+		// eventually terminate, so we're generous with our retry policy.
+		if len(fileSpecs) > numNeedUploading/2 {
+			return fmt.Errorf(
+				"most remaining uploads (%d/%d) have failed, giving up",
+				len(fileSpecs), numNeedUploading,
 			)
-			if err != nil {
-				return err
-			}
-			if len(fileSpecsBatch) != len(response.CreateArtifactFiles.Files.Edges) {
-				return fmt.Errorf(
-					"expected %v upload URLs, got %v",
-					len(fileSpecsBatch),
-					len(response.CreateArtifactFiles.Files.Edges),
-				)
-			}
-			// Save birth artifact ids, schedule uploads.
-			for i, edge := range response.CreateArtifactFiles.Files.Edges {
-				name := fileSpecsBatch[i].Name
-				entry := manifest.Contents[name]
-				entry.BirthArtifactID = &edge.Node.Artifact.Id
-				manifest.Contents[name] = entry
-				if edge.Node.UploadUrl == nil {
-					numDone++
-					continue
-				}
-				numInProgress++
-				task := &filetransfer.Task{
-					FileKind: filetransfer.RunFileKindArtifact,
-					Type:     filetransfer.UploadTask,
-					Path:     *entry.LocalPath,
-					Url:      *edge.Node.UploadUrl,
-					Headers:  edge.Node.UploadHeaders,
-				}
-				task.SetCompletionCallback(
-					func(t *filetransfer.Task) {
-						taskResultsChan <- TaskResult{t, name}
-					},
-				)
-				as.FileTransferManager.AddTask(task)
-			}
 		}
-		// Wait for filetransfer to catch up. If there's nothing more to schedule, wait for all in progress tasks.
-		for numInProgress > maxBacklog || (len(fileSpecsBatch) == 0 && numInProgress > 0) {
-			numInProgress--
-			result := <-taskResultsChan
-			if result.Task.Err != nil {
-				// We want to retry when the signed URL expires. However, distinguishing that error from others is not
-				// trivial. As a heuristic, we retry if the request failed more than an hour after we fetched the URL.
-				if time.Since(nameToScheduledTime[result.Name]) < 1*time.Hour {
-					return result.Task.Err
-				}
-				delete(nameToScheduledTime, result.Name) // retry
+		if len(fileSpecs) > 0 {
+			slog.Warn("some files failed to upload, retrying", "count", len(fileSpecs))
+		}
+	}
+	return nil
+}
+
+func (as *ArtifactSaver) processFiles(
+	manifest *Manifest, fileSpecs map[string]gql.CreateArtifactFileSpecInput,
+) (map[string]gql.CreateArtifactFileSpecInput, error) {
+
+	// Channels to get responses from the batch url retrievers.
+	readyChan := make(chan serverFileResponse, as.batchSize)
+	errorChan := make(chan error)
+
+	doneChan := make(chan *filetransfer.Task, as.batchSize)
+	mustRetry := map[string]gql.CreateArtifactFileSpecInput{}
+
+	numTotal := len(fileSpecs)
+	numActive := 0
+	numDone := 0
+
+	fileSpecBatches := batchify(fileSpecs, as.batchSize)
+
+	for numDone+len(mustRetry) < numTotal {
+		// If we have less than a batch worth of requests open, start a new batch.
+		if numActive <= as.batchSize && len(fileSpecBatches) > 0 {
```

Comment:
should this entire func just be a gorountine outside of the main select loop? My brain is hurting trying to reason about the select loop blocking the batch retrieving

Reply:
That's how I had it at first (in the other PR you reviewed for this). That works, it just requires atomic reads/writes on `numActive`. In some ways it's nicer. This style here does actually slightly slow down getting more batches. I moved it out of a goroutine because I thought it was actually easier to reason about and I decided to prioritize comprehensibility for this PR, but if this form feels more confusing to you than having a separate goroutine I think I'll move it back to one.